src/transformers/models/kosmos2/modeling_kosmos2.py

```diff
@@ -0,0 +1,2106 @@
+            if module.out_proj.bias is not None:
+                module.out_proj.bias.data.zero_()
+        elif isinstance(module, Kosmos2TextFFN):
+            nn.init.normal_(module.fc1.weight, std=std)
+            nn.init.normal_(module.fc2.weight, std=std)
+            if module.fc1.bias is not None:
+                module.fc1.bias.data.zero_()
+            if module.fc2.bias is not None:
+                module.fc2.bias.data.zero_()
+        elif isinstance(module, Kosmos2TextForCausalLM):
+            nn.init.normal_(module.lm_head.weight, std=std)
+            if module.lm_head.bias is not None:
+                module.lm_head.bias.data.zero_()
+        elif isinstance(module, Kosmos2ImageToTextConnector):
+            nn.init.normal_(module.dense.weight, std=std)
+            if module.dense.bias is not None:
+                module.dense.bias.data.zero_()
+        elif isinstance(module, Kosmos2TextTransformer):
+            module.embed_tokens.weight.data.normal_(mean=0.0, std=std)
+            if module.embed_tokens.padding_idx is not None:
+                module.embed_tokens.weight.data[module.embed_tokens.padding_idx].zero_()
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, (Kosmos2VisionEncoder, Kosmos2TextTransformer)):
+            module.gradient_checkpointing = value
+
+
+class Kosmos2VisionModel(Kosmos2PreTrainedModel):
+    config_class = Kosmos2VisionConfig
+    main_input_name = "pixel_values"
+
+    # Copied from transformers.models.clip.modeling_clip.CLIPVisionModel.__init__ with CLIP_VISION->KOSMOS2_VISION,CLIP->Kosmos2,self.vision_model->self.model
+    def __init__(self, config: Kosmos2VisionConfig):
+        super().__init__(config)
+        self.model = Kosmos2VisionTransformer(config)
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    # Copied from transformers.models.clip.modeling_clip.CLIPVisionModel.get_input_embeddings with CLIP_VISION->KOSMOS2_VISION,CLIP->Kosmos2,self.vision_model->self.model
+    def get_input_embeddings(self) -> nn.Module:
+        return self.model.embeddings.patch_embedding
+
+    @add_start_docstrings_to_model_forward(KOSMOS2_VISION_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=BaseModelOutputWithPooling, config_class=Kosmos2VisionConfig)
+    def forward(
+        self,
+        pixel_values: Optional[torch.FloatTensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, BaseModelOutputWithPooling]:
+        r"""
+        Returns:
+
+        """
+        return self.model(
+            pixel_values=pixel_values,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+
+class Kosmos2TextModel(Kosmos2PreTrainedModel):
```

Comment:
I don't think we need this additional layer. We could just have a Kosmos2Decoder and a Kosmos2Encoder. 
Instead of having 2 extra layers `Kosmos2TextModel` is equivalent to `Kosmos2TextTransformer`.
Same for the vision model.

Reply:
FYI this was added by @patil-suraj for CLIP so that people can also load only the vision or text model.

If we decide to remove this class, then people can't load the individual vision/text encoders, cause `from_pretrained` will complain about the keys

Reply:
With the current implementation, I make sure loading a `Kosmos2ForConditionalGeneration` checkpoint to `Kosmos2Model` works. But it won't work if we load `Kosmos2ForConditionalGeneration` checkpoint to text or vision component.

I indeed take the idea from CLIP. However, CLIP is slightly simpler than this text-vision conditional generation model. There is a way to make loading to component work, but it will require another arrangement of internal layer structure.

Reply:
Ok, for KOSMOS-2 people probably don't want to load the individual vision/text components, so that's fine.

For CLIP, having the ability to load either the text or vision encoder is pretty handy